compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -246,11 +249,125 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
+                struct BorrowVisitor<'a, 'tcx> {
+                    found: bool,
+                    thir: &'a Thir<'tcx>,
+                    tcx: TyCtxt<'tcx>,
+                }
+
+                impl<'a, 'tcx> Visitor<'a, 'tcx> for BorrowVisitor<'a, 'tcx> {
+                    fn thir(&self) -> &'a Thir<'tcx> {
+                        self.thir
+                    }
+
+                    fn visit_expr(&mut self, expr: &Expr<'tcx>) {
+                        match expr.kind {
+                            ExprKind::Field { lhs, .. } => {
+                                if let ty::Adt(adt_def, _) = self.thir[lhs].ty.kind() {
+                                    if (Bound::Unbounded, Bound::Unbounded)
+                                        != self.tcx.layout_scalar_valid_range(adt_def.did)
+                                    {
+                                        self.found = true;
+                                    }
+                                }
+                                visit::walk_expr(self, expr);
+                            }
+                            ExprKind::Deref { .. } => {}
+                            _ => {
+                                visit::walk_expr(self, expr);
+                            }
+                        }
+                    }
+                }
+
+                let mut visitor = BorrowVisitor { found: false, thir: self.thir, tcx: self.tcx };
+                visit::walk_expr(&mut visitor, expr);
+
+                if visitor.found {
+                    match borrow_kind {
+                        BorrowKind::Shallow | BorrowKind::Shared | BorrowKind::Unique
+                            if !self.thir[arg]
+                                .ty
+                                .is_freeze(self.tcx.at(self.thir[arg].span), self.param_env) =>
+                        {
+                            self.requires_unsafe(expr.span, BorrowOfLayoutConstrainedField);
+                        }
+                        BorrowKind::Mut { .. } => {
+                            self.requires_unsafe(expr.span, MutationOfLayoutConstrainedField);
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            ExprKind::Assign { lhs, .. } | ExprKind::AssignOp { lhs, .. } => {
+                let lhs = match self.thir[lhs].kind {
+                    ExprKind::Scope { value, .. } => value,
+                    _ => lhs,
+                };
+                if let ExprKind::Field { lhs, .. } = self.thir[lhs].kind {
+                    if let ty::Adt(adt_def, _) = self.thir[lhs].ty.kind() {
+                        if (Bound::Unbounded, Bound::Unbounded)
+                            != self.tcx.layout_scalar_valid_range(adt_def.did)
+                        {
+                            self.requires_unsafe(expr.span, MutationOfLayoutConstrainedField);
+                        }
+                    }
+                }
```

Comment:
I'd expect the same logic for assignments as for borrows, i.e. look for layout-constrained fields inside the place we assign to. Here's an example:

```rs
let mut foo = unsafe { NonZero((1,)) };
foo.0.0 = 0;
```

This is caught by the MIR unsafeck, but not by this implementation that just looks one level deep.

Reply:
Fixed. I have also added test cases for all of the examples that you've mentioned. Thanks for your patience in (re-)reviewing this!

Reply:
I'm the one who should thank you! The reason we need so many rounds of review is because I could only catch problems one by one, so thank you for bearing my inexperience with reviews :sweat_smile: